### PR TITLE
vere: mars/urth

### PR DESCRIPTION
**Pull request by joemfb**
The PR moves the event log into the worker process (along with the boot process and event replay), thereby redefining the roles (and IPC) in our multiprocess architecture (from "king" and "serf" to "urth" and "mars"). This work is a prerequisite for event log truncation and a variety of improvements to the replay process (both to be realized by collocating the event log and snapshot management).

Still to be done:

- CI is broken due to ~~#5537~~ (*edit*: #5317), partially fixed by #5470
- lockfile management is broken, partially fixed in #5575 
- `urbit-worker` commandline args are messy and incomplete
- the serf module (`worker/serf.c`, plus its entrypoint in `worker/main.c`) needs to be deleted

### CODE REVIEW

File: pkg/urbit/worker/mars.c [hunk 1]
```diff
@@ -0,0 +1,1159 @@
...
+    if ( c3y == u3v_poke_sure(mil_w, u3k(*eve), &pro) ) {
+      u3z(dud);
+      *out = pro;
+      return c3y;
+    }
+    else {
+      *out = u3nt(dud, pro, u3_nul);
+      return c3n;
+    }
+  }
```
Comment (mcevoypeter) [inline pkg/urbit/worker/mars.c:319]: Same "get rid of the `else`" comment.

File: pkg/urbit/worker/mars.c [hunk 2]
```diff
@@ -0,0 +1,1159 @@
...
+
+  if ( c3n == ret_o ) {
+    //  XX better
+    //
+    fprintf(stderr, "mars: commit fail\r\n");
+    exit(1);
+  }
+  else {
+    _mars_flush(mar_u);
+  }
```
Comment (mcevoypeter) [inline pkg/urbit/worker/mars.c:618]: Same "drop the `else`" comment.

File: pkg/urbit/worker/mars.c [hunk 3]
```diff
@@ -0,0 +1,1159 @@
...
+    u3z(_mars_sure_feck(mar_u, pre_w, vir));
+    return c3y;
+  }
+
+  //  XX produce/print trace, reclaim on meme, retry on %intr, &c
+  //
+  // {
+  //   u3_noun mot, tan;
+  //   u3x_cell(vir, &mot, &tan);
+  // }
```
Comment (mcevoypeter) [inline pkg/urbit/worker/mars.c:639]: Can we remove this?

**Review by mcevoypeter — COMMENTED** [review]
Lots of small comments that can be broadly bucketed into a few different categories:
- "Let's create a separate header file for this module." 
- "Let's delete this commented-out block of code if we can."
- "Let's remove this `else` statement."
- "Let's use a macro in place of a literal number here."

File: pkg/urbit/noun/allocate.c [hunk 4]
```diff
@@ -1964,6 +1963,122 @@ u3a_maid(FILE* fil_u, c3_c* cap_c, c3_w wor_w)
...
+          box_u->eus_w -= 1;
+        }
+#else
+        if ( -1 == (c3_w)box_u->use_w ) {
+          box_u->use_w = 0x80000000;
+        }
+        else {
+          box_u->use_w += 1;
+        }
+#endif
```
[the remark this answers is not in the document]
Comment (joemfb) [reply pkg/urbit/noun/allocate.c:2055]: This is very old code in the GC implementation behind `|mass`, I'm simply copying it out of `worker/serf.c` (which will be shortly deleted). I'm in favor of refactoring it, but not piecemeal (there's quite a big more code implicated in `allocate.c`).

File: pkg/urbit/noun/allocate.c [hunk 5]
```diff
@@ -1964,6 +1963,122 @@ u3a_maid(FILE* fil_u, c3_c* cap_c, c3_w wor_w)
...
+        }
+        else {
+          box_u->eus_w -= 1;
+        }
+#else
+        if ( -1 == (c3_w)box_u->use_w ) {
+          box_u->use_w = 0x80000000;
+        }
+        else {
+          box_u->use_w += 1;
```
[the remark this answers is not in the document]
Comment (joemfb) [reply pkg/urbit/noun/allocate.c:2053]: moar GC impl

File: pkg/urbit/vere/db/lmdb.c [hunk 6]
```diff
@@ -217,6 +217,112 @@ u3_lmdb_gulf(MDB_env* env_u, c3_d* low_d, c3_d* hig_d)
...
+  c3_w    ops_w, ret_w;
+
+  itr_u->red_o = c3n;
+  itr_u->nex_d = nex_d;
+  itr_u->las_d = las_d;
+
+  //  create a read-only transaction.
+  //
+  ops_w = MDB_RDONLY;
+  if ( (ret_w = mdb_txn_begin(env_u, 0, ops_w, &itr_u->txn_u)) ) {
```
[the remark this answers is not in the document]
Comment (joemfb) [reply pkg/urbit/vere/db/lmdb.c:241]: 80 columns without multiple lines ... ;)

File: pkg/urbit/vere/disk.c [hunk 7]
```diff
@@ -181,284 +148,308 @@ _disk_serialize_v1(u3_fact* tac_u, c3_y** out_y)
...
+           ^ (dat_y[3] << 24);
 
 #ifdef DISK_TRACE_CUE
     u3t_event_trace("king disk cue", 'B');
 #endif
 
     //  XX u3m_soft?
     //
-    job = u3ke_cue(u3i_bytes(val_i - 4, dat_y + 4));
+    *job = u3ke_cue(u3i_bytes(len_i - 4, dat_y + 4));
```
[the remark this answers is not in the document]
Comment (joemfb) [reply pkg/urbit/vere/disk.c]: IMO, `sizeof` is appropriate for abstracting things that may change. But mug is defined to be a 31-bit, nonzero has, so abstracting over its size just seems obfuscatory. And there's an XOR/shift expression right above this with explicit `0-3` array indexing ...

File: pkg/urbit/worker/main.c [hunk 8]
```diff
@@ -451,6 +455,306 @@ _cw_pack(c3_i argc, c3_c* argv[])
...
+    fprintf(stderr, "boot: fail\r\n");
+    exit(1);
+  }
+  else {
+    fprintf(stderr, "boot: good\r\n");
+    exit(0);
+  }
+
+  return c3y;
+}
```
[the remark this answers is not in the document]
Comment (joemfb) [reply pkg/urbit/worker/main.c:482]: This return type is dictated by the IPC implementation in `newt.c`. (Hopefully, we'll be replacing that soon.)

File: pkg/urbit/worker/mars.c [hunk 9]
```diff
@@ -0,0 +1,1159 @@
...
+  //    XX maybe use 2^23 (~16 MB) and 2^26 (~268 MB?
+  //
+  //    XX these thresholds should trigger notifications sent to the king
+  //    instead of directly triggering these remedial actions.
+  //
+  {
+    u3_noun pri = u3_none;
+    c3_w pos_w = u3a_open(u3R);
+    c3_w low_w = (1 << 27);
+    c3_w hig_w = (1 << 22);
```
[the remark this answers is not in the document]
Comment (joemfb) [reply pkg/urbit/worker/mars.c:253]: These would be handy. But I'm not sure where to put them -- I don't think we want unprefixed symbols in our public headers, but I also don't love `u3_KiB` ... `-impl.h` headers might be nice for this.

File: pkg/urbit/include/vere/vere.h [hunk 10]
```diff
@@ -923,7 +897,24 @@
...
+                     c3_y**   out_y);
+
+      /* u3_disk_sift(): parse a persisted event buffer.
+      */
+        c3_o
+        u3_disk_sift(u3_disk* log_u,
+                     size_t   len_i,
+                     c3_y*    dat_y,
+                     c3_l*    mug_l,
+                     u3_noun*   job);
```
[the remark this answers is not in the document]
Comment (mcevoypeter) [reply pkg/urbit/include/vere/vere.h]: Rescinding this comment as the `disk` module will soon be replaced by the `epoc` module in the epoch-based event log implementation.

**Comment by joemfb** [comment]
My apologies for the force-push -- getting these merges right took more attempts than I expected.

**Comment by joemfb** [comment]
@Fang- PTAL at [58be26e](https://github.com/urbit/urbit/pull/5596/commits/58be26eba8d7822a114b3217e8fc6f15b683e936) and [863fd10](https://github.com/urbit/urbit/pull/5596/commits/863fd10a493726a60c91a4040fd63ecca500a773)

**Pull request merged**